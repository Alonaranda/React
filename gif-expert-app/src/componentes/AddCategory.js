import React, {useState} from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setValue] = useState('');

    const handleInputText = (e) => {
        setValue(e.target.value);
        //console.log('Change', inputValue);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        //console.log('Submit!', inputValue);
        if(inputValue.trim().length > 2){
            setCategories(cate => [inputValue, ...cate, ]);
            setValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputText}
                placeholder='Buscado tu Gif'
            ></input>
        </form>
        
    )
}

//PropTypes
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
