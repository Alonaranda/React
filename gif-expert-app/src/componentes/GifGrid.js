import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from 'prop-types';
//import { getGifs } from "../helpers/getGifs";


export const GifGrid = ({ category }) => {
  //Al ejecutar una funcion dentro de una api, se vuelve un ciclo infinito
  //UseEffect hace que no hagamos ciclo infinito
  //Ejecuta cierto codigo de manera condicional
  //El arreglo vacio, indica que solo se ejecuta una vez

  
    
  //const [images, setImages] = useState([]);
    //Usamos helpers
    const {data:images, loading} = useFetchGifs(category);
    //console.log(data, loading);

    /*
    useEffect(() => {
        getGifs(category)
            .then(imgs =>{
                setImages(imgs);
        })
    }, [category]);//Si la category cambia, se vuelve a ejecutar
    */
  

  return (
    <>
      <h3 className="animate_animated animate_fadeIn">{category}</h3>
      {loading && <p className="animate_animated animate_flash">Loading...</p>}

      <div className="card-grid">
        {images.map(img => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}
