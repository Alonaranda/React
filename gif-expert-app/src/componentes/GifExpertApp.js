import React, {useState} from 'react'
import { AddCategory } from './AddCategory';
import { GifGrid } from './GifGrid';

export const GifExpertApp = ({defaultCategories=[]}) => {
    //const categories = ['Inuyasha', 'Naruto', 'Samurai X'];
    const [categories, setCategories] = useState(defaultCategories);
    /*
    const handleAdd = () => {
        //setCategories([...categories, 'DeathNote']);
        setCategories( cate => [...cate, 'DeathNote']);
    }*/

    //Para pasar datos en el <AddCategory/> agregamos que vamos a utilizar
    return (
        <>
            <h2>GifExpertApp</h2>
             
            <AddCategory setCategories={setCategories}/>
            <hr/>
            <ol>
                {
                    categories.map(category => 
                        <GifGrid 
                        key={category}
                        category={category}
                        />
                    //<li key={category}>{category}</li>
                    ) //El key no debe ser el indice, casi siempre es el id de la BD
                }
            </ol>
        </>
    )
}

