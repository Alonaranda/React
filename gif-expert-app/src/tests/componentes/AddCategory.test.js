import React from 'react';
import '@testing-library/jest-dom';
import { AddCategory } from "../../componentes/AddCategory";
import {shallow} from 'enzyme';


describe('Probando AddCategory', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    });

    //Requiere la funcion SetCategories como prop
    test('Snapshot de <AddCategory/>', () => {
        expect(wrapper).toMatchSnapshot();
    });

    //Simular cambios en input
    test('Evaluar cambias en input', () => {
        const input = wrapper.find('input');
        const value = 'hola';
        
        input.simulate('change', {
            target: {value}
        });
        expect(wrapper.find('p').text().trim()).toBe(value);
    });


    //Simular submit, reinicia test, no se llama funcion
    test('Prevent default', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });
        expect(setCategories).not.toHaveBeenCalled();
    });

    //Se limpia el input
    test('Llamar setCategories y limpiarla ', () => {
        //console.log('limpiar');
        //Change
        const input = wrapper.find('input');
        const value = 'hola';
        input.simulate('change', {
            target: {value}
        });
        //Submit
        wrapper.find('form').simulate('submit', {
            preventDefault(){},
        });
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        expect(wrapper.find('input').prop('value')).toBe('');

    });
});