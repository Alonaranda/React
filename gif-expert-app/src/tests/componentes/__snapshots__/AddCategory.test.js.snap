// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando AddCategory Snapshot de <AddCategory/> 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Buscado tu Gif"
    type="text"
    value=""
  />
</form>
`;
